composer/core/evaluator.py

```diff
@@ -51,17 +48,25 @@ def should_eval(state: State, event: Event):
         if eval_at_fit_end and event == Event.FIT_END and state.timestamp.batch != last_batch_seen:
             return True
 
-        if eval_interval.unit == TimeUnit.EPOCH and int(
-                state.timestamp.epoch) % int(eval_interval) == 0 and event == Event.EPOCH_END:
+        assert state.previous_timestamp is not None
+        if eval_interval.unit in {TimeUnit.EPOCH, TimeUnit.BATCH, TimeUnit.TOKEN, TimeUnit.SAMPLE}:
+            previous_count = state.previous_timestamp.get(eval_interval.unit)
+            count = state.timestamp.get(eval_interval.unit)
+        else:
+            assert state.max_duration is not None
+            previous_count = state.previous_timestamp.get(state.max_duration.unit)
+            count = state.timestamp.get(state.max_duration.unit)
+
+        threshold_passed = math.floor(previous_count / eval_interval.value) != math.floor(count / eval_interval.value)
+
+        if eval_interval.unit == TimeUnit.EPOCH and event == Event.EPOCH_END and threshold_passed:
```

Comment:
delete last_batch_seen? because of prev timestamp

Reply:
this logic was here before and is slightly different. It basically says "Checkpoint at fit end if you didn't already checkpoint that batch". but the prev timestamp doesn't track when you last checkpointed, so doesn't cover this. i'm sure there is a nicer way to do this, but i think out of scope of this pr.